### PR TITLE
Update checks

### LINKED ISSUE #700: Split CI tests
### Feature Request

The workflow runs started to take a bit longer.
We could use some inspiration from [actions in protostar](https://github.com/software-mansion/protostar/actions) to speed up the process.

### PR DESCRIPTION
**Pull request by war-in**
Closes #700 

## Introduced changes

This PR splits CI tests into:
- installing dependencies 
- checking format etc
- unit & e2e tests
- docs tests

The last three are running in parallel 🔥 

Comparison:
- before
image

- after
image

##

- [ ] This PR contains breaking changes

### CODE REVIEW

File: .github/workflows/checks.yml [hunk 1]
```diff
@@ -8,49 +8,108 @@ on:
...
+      matrix:
+        python-version: [ "3.8", "3.9" ]
+    steps:
+      - uses: actions/checkout@v3
+        with:
+          submodules: recursive
+      - name: Set up Python ${{ matrix.python-version }}
+        uses: actions/setup-python@v4
+        with:
+          python-version: ${{ matrix.python-version }}
```
Comment (drknzz) [inline .github/workflows/checks.yml:56]: Why do we set up python in every job? Is setting it in `setup` not enough?

Comment (war-in) [reply .github/workflows/checks.yml]: Nope, it has to be set up for every job

File: .github/workflows/checks.yml [hunk 2]
```diff
@@ -8,49 +8,128 @@ on:
...
+    needs: setup
+    strategy:
+      matrix:
+        python-version: [ "3.8", "3.9" ]
+    steps:
+      - uses: actions/checkout@v3
+        with:
+          submodules: recursive
+      - name: Install poetry
+        run: |
```
Comment (cptartur) [inline .github/workflows/checks.yml:46]: Remove newline

File: .github/workflows/checks.yml [hunk 3]
```diff
@@ -8,49 +8,128 @@ on:
...
+    runs-on: ubuntu-latest
+    needs: setup
+    strategy:
+      matrix:
+        python-version: [ "3.8", "3.9" ]
+    steps:
+      - uses: actions/checkout@v3
+        with:
+          submodules: recursive
+      - name: Install poetry
```
Comment (cptartur) [inline .github/workflows/checks.yml:45]: Maybe a comment why we need to reinstall poetry even though we are using cache?

File: .github/workflows/checks.yml [hunk 4]
```diff
@@ -8,49 +8,128 @@ on:
...
+        run: |
+          python -m pip install --upgrade pip
+          pip install poetry
+      - name: Set up Python ${{ matrix.python-version }}
+        uses: actions/setup-python@v4
+        with:
+          python-version: ${{ matrix.python-version }}
+          cache: 'poetry'
+
```
Comment (cptartur) [inline .github/workflows/checks.yml:124]: unnecessary newline

**Review by cptartur — APPROVED** [review]

**Review by drknzz — APPROVED** [review]
Great. I think we might need to change the required checks though?

**Pull request merged**